src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs

```diff
@@ -3381,10 +3381,23 @@ private BoundExpression BindNullCoalescingOperator(BinaryExpressionSyntax node,
                 return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
             }
 
-            // SPEC: Otherwise, if A exists and is a non-nullable value type, a compile-time error occurs.
-            if ((object)optLeftType != null && optLeftType.IsValueType && !optLeftType.IsNullableType())
+            // SPEC: Otherwise, if A exists and is a non-nullable value type, a compile-time error occurs. First we check for the pre-C# 8.0
+            // SPEC: condition, to ensure that we don't allow previously illegal code in old language versions.
+            if ((object)optLeftType != null && !optLeftType.IsReferenceType && !isLeftNullable)
             {
-                return GenerateNullCoalescingBadBinaryOpsError(node, leftOperand, rightOperand, Conversion.NoConversion, diagnostics);
+                // Prior to C# 8.0, the spec said that the left type must be either a reference type or a nullable value type. This was relaxed
+                // with C# 8.0, so if the feature is not enabled then issue a diagnostic and return
+                if (!optLeftType.IsValueType && !isLeftNullable)
+                {
+                    if (!CheckFeatureAvailability(node, MessageID.IDS_FeatureUnconstrainedTypeParameterInNullCoalescingOperator, diagnostics))
+                    {
+                        return new BoundNullCoalescingOperator(node, leftOperand, rightOperand, Conversion.NoConversion, CreateErrorType());
```

Comment:
>return new BoundNullCoalescingOperator(node, leftOperand, rightOperand, Conversion.NoConversion, CreateErrorType()); [](start = 24, length = 116)

We should report diagnostics and keep binding normally. In general, language version shouldn't affect how we parse\bind things apart from diagnostics.

Reply:
I'm attempting to replicate what happened prior to C# 8.0 for unconstrained type parameters, which was to simply report an error on this node and stop binding.

---
In reply to: [217204494](https://github.com/dotnet/roslyn/pull/29724#discussion_r217204494) [](ancestors = 217204494)

Reply:
> I'm attempting to replicate what happened prior to C# 8.0 for unconstrained type parameters, which was to simply report an error on this node and stop binding.

We don't do this. Language version should not have an effect like this. The only effect it can have is to report diagnostics

---
In reply to: [217216299](https://github.com/dotnet/roslyn/pull/29724#discussion_r217216299) [](ancestors = 217216299,217204494)